### ISSUE TITLE
OS/X Metal renderer

### ISSUE THREAD

**Issue by christianparpart** [contour-terminal/contour#310]
Evaluate possibility to render via Metal on OS/X instead of OpenGL (or software).

* https://developer.apple.com/metal/
* https://github.com/msorvig/qt-macos-examples/tree/dev/metalwidget

**Comment by WSLUser** [comment]
Maybe Vulkan would be a better alternative. At least that rendering engine works on all OS

**Comment by WSLUser** [comment]
Seems Vulkan has already accounted for Metal and probably harnesses the same APIs: https://www.vulkan.org/portability

**Comment by WSLUser** [comment]
See https://github.com/KhronosGroup/Vulkan-Hpp for getting started if you choose to pursue this.

**Comment by christianparpart** [comment]
> See https://github.com/KhronosGroup/Vulkan-Hpp for getting started if you choose to pursue this.

Vulkan as much as DirectX as much as Metal. All of them. But no tickets.

I created this ticked as a slight wink towards @uspasojevic96 as he recently said he can only help with OSX stuff. If one of you is willing to support me in adding a Vulkan renderer, i would be more than happy :-)

**Comment by WSLUser** [comment]
If you do end up supporting multiple renderers, how do you choose which one to use for a user install? Perf, features? Something in between? What if a user wants  to choose a specific one, will there be separate builds created based on the rendering engine?

**Comment by uspasojevic96** [comment]
> > See https://github.com/KhronosGroup/Vulkan-Hpp for getting started if you choose to pursue this.
> 
> 
> 
> Vulkan as much as DirectX as much as Metal. All of them. But no tickets.
> 
> 
> 
> I created this ticked as a slight wink towards @uspasojevic96 as he recently said he can only help with OSX stuff. If one of you is willing to support me in adding a Vulkan renderer, i would be more than happy :-)

I am more than happy to give it a shot, given it will take bit of time since I also said I did bare minimum with Metal (:

**Comment by christianparpart** [comment]
@WSLUser based on platform, their official or main rendering pipeline or alternatively the most stable one.

Wrt features, the renderer ONLY renders basic primitive 2D textures into a 2D surface.
There is nothing fancy exceptasbe being faster than another.

In the end the user should always b abl to override per config flag.

@uspasojevic96 no hurries man. :)

**Comment by WSLUser** [comment]
> the renderer ONLY renders basic primitive 2D textures into a 2D surface.

I'm still waiting for 3d Animations to be a thing for terminals. Feel free to be the first (that I'd be aware of anyways). I don't think even Kitty is doing that yet but that work would likely require extending something out of Kitty's graphic protocol and not Sixel and of course there's no rush for that, especially with all the discussion going around in the terminal world surrounding Sixel at this time.

**Comment by christianparpart** [comment]
> especially with all the discussion going around in the terminal world surrounding Sixel at this time

sixel (and image in general) are a too sensitive topic in TE world it seems. way too many strong opinions.

> 3d animactions

tbh, you can always transform your 2d terminal plane into 3d space and have fancy effects for switching between terminal tabs, oh yes. but that's just special effects (actually, I'm having a label for that, and had a FEW nice tricks in my head) :dagger: 

> I don't think even Kitty is doing that

Kitty's graphics protocol is heavy and slow. But indeed it does support placing images into arbitrary z-axis planes, which comes with its own problems (albeit benefits do exist, too). I personally think real 3d is best left at awesome games in the GUI using GL etc. :D

**Comment by WSLUser** [comment]
> I personally think real 3d is best left at awesome games in the GUI using GL etc. :D

Yeah I get that. I just like thinking how we can push the boundaries of what you can do with a terminal and actually run graphic intensive stuff through a terminal on a headless host. Not that it's needed but more of an exploration/fun way to utilize the terminal and who knows? Maybe it could have some real use-cases from a prod perspective.

**Comment by uspasojevic96** [comment]
> > I personally think real 3d is best left at awesome games in the GUI using GL etc. :D
> 
> Yeah I get that. I just like thinking how we can push the boundaries of what you can do with a terminal and actually run graphic intensive stuff through a terminal on a headless host. Not that it's needed but more of an exploration/fun way to utilize the terminal and who knows? Maybe it could have some real use-cases from a prod perspective.

It's supposed to be terminal emulation, if it isn't in spec I don't see why would people do that in production, at least that's my point of view

**Comment by christianparpart** [comment]
We have plenty extensions. But for your wish there does exist something already

**Comment by christianparpart** [comment]
We have plenty extensions. But for your wish there does exist something already. You can run the X server over TCP/IP. :)

It does not make sense to recreate that.
